### ISSUE TITLE
Don't schedule driver to run on low priority instances

### ISSUE THREAD

**Issue by jaley** [Azure/aztk#527]
I mentioned in #375 that it makes sense for us to run the driver on the master instance, because the master has capacity and won't be doing much else.

I noticed that currently, the driver may be scheduled to run on a low priority instance, which is causing stability problems on our ETL. Though I mentioned in the other ticket that I don't think it makes good sense to run the driver in client mode on an instance that's also running a worker, I think it's actually broken to schedule the driver to run on low priority instances when regular instances exist in the cluster.

The behaviour right now, if the instance is pre-empted, is that the whole job fails. For a big job, it's basically just as bad as losing the master, which we already have set to prefer non-preemptible instances.

**Comment by timotheeguerin** [comment]
We can add functionality that it should run on a dedicated(or any node you'd want) as long as there is a available spot(From max task per node restriction on Batch which is x4 the number of cores of a node)
This should be a minor change, for the run on master to guarantee this is going to work all the time it is a bigger item which we still need to think about. Probably needs some options on where you want the driver to run (master, not master, dedicated, etc.)

**Comment by jaley** [comment]
Yeah, this is the biggest stability issue we have right now. Scheduling the driver to run on a machine that's busy running workers causes big problems.

Several hours and 50TB or so into a large ETL job, the driver dies and the job restarts with a whole new retry from the beginning. I check the driver logs and see errors that a task has failed 4 times so it's giving up. Check the executor logs where the failing task was running, and I see something like this:

```
2018-04-30 16:06:55 WARN  Executor:87 - Issue communicating with driver in heartbeater
org.apache.spark.rpc.RpcTimeoutException: Futures timed out after [10 seconds]. This timeout is controlled by spark.executor.heartbeatInterval
	at org.apache.spark.rpc.RpcTimeout.org$apache$spark$rpc$RpcTimeout$$createRpcTimeoutException(RpcTimeout.scala:47)
	at org.apache.spark.rpc.RpcTimeout$$anonfun$addMessageIfTimeout$1.applyOrElse(RpcTimeout.scala:62)
	at org.apache.spark.rpc.RpcTimeout$$anonfun$addMessageIfTimeout$1.applyOrElse(RpcTimeout.scala:58)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:76)
	at org.apache.spark.rpc.RpcEndpointRef.askSync(RpcEndpointRef.scala:92)
	at org.apache.spark.executor.Executor.org$apache$spark$executor$Executor$$reportHeartBeat(Executor.scala:785)
	at org.apache.spark.executor.Executor$$anon$2$$anonfun$run$1.apply$mcV$sp(Executor.scala:814)
	at org.apache.spark.executor.Executor$$anon$2$$anonfun$run$1.apply(Executor.scala:814)
	at org.apache.spark.executor.Executor$$anon$2$$anonfun$run$1.apply(Executor.scala:814)
	at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1988)
	at org.apache.spark.executor.Executor$$anon$2.run(Executor.scala:814)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.TimeoutException: Futures timed out after [10 seconds]
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:201)
	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:75)
	... 14 more
```

So, the driver and the executor are kind of pointing at each other. The problem is that the driver machine is too busy to respond to incoming network requests from the many hundreds of executors in the cluster, because it's also running workers.

I'll get *some* mileage out of increasing timeouts and hoping for the best, but it's still going to be flaky. I can also scale back the workers, so that we don't claim all CPUs, memory, etc on all machines, but the problem is I can only make that change to *all* workers in preparation for any one of them getting picked at random to run the driver.

As I mentioned in the above ticket, I think the sensible options here are:

* Run the driver (in *client mode*) and spark master on the same (dedicated) instance, with no workers there.
* Pick a node at random to run the driver, but use *cluster mode* so that it exits immediately, letting the Spark master pick a node to run it, subtracting driver resources from the worker chosen.
  * I expect it'll be more difficult to ensure that the driver isn't running on a low priority node this way, as it's spark's choice where the driver runs, and spark has no visibility of that.

**Comment by timotheeguerin** [comment]
Yeah so I looked at the code, and we are actually already picking the master node as the target for running the Batch task triggering `spark-submit` but it seems we are not running spark in `client` mode. Which I would guess is the problem.

**Comment by timotheeguerin** [comment]
Hhm, so `--deploy-mode client` is the default for spark apparently. So I would then expect it to work as expected(Schedule driver on master). Did you submit multiple application at the same time? maxTaskPerNode is still set to 1 it seems which could explain why the first application got scheduled on the master and the following on another node

**Comment by jaley** [comment]
Client mode is the default, yes, so that'll mean the Driver JVM runs on the instance that calls `spark-submit` unless you explicit ask for cluster mode.

I'm definitely seeing the driver running on worker machines though. I checked in `/mnt/batch/tasks/workitems/` and can see that the Batch task appears to have been scheduled to a node at random, which on my clusters is usually a low priority node that also runs executors, because 90% of my cluster is low priority.

**Comment by timotheeguerin** [comment]
Hhm, which version of aztk are you using?
Did you run multiple application at the same time?

Also can you use BatchLabs or the portal to check the job(Same name as the cluster id) if you havnen't deleted your cluster and see if the task you have have the right affinityId(Should be the id of the master node) Also check the execution information to see if the node that got picked is the same

**Comment by jaley** [comment]
We're running from master, as we need some features not yet released. I *think* the revision was 0015e22d when we last updated our image.

We don't run multiple applications on the same cluster. We much prefer to run applications on dedicated clusters, given we can easily create as many as we need for as long as we need them. This is why, for us, running the driver on the master instance is preferable -- we can be sure there will only be one driver.

My clusters from yesterday have been torn down, but actually I just checked the ones I have online today and you're completely right! The task affinity appears to be set, and the drivers are in fact running on the master instance for all jobs I have online right now. So, I have no idea why that wasn't the case yesterday, but it seems to be fine now.

When was this introduced? If this is now the designed behaviour, we might as well close #352 and I'll just monitor my jobs to make sure the driver is always on the master from now on?

**Comment by timotheeguerin** [comment]
This feature has been added since 0.6.0 release at least. After testing a bit I believe that the affinity feature of Azure Batch has some issues(That we have reported to the service). It might not always respect what you give even if there is an available spot. In the meantime I am not 100% sure of what we could do.
On way would be to fail the task right away if the node picked is not dedicated which would require you to try resubmitting until you get a dedicated node.
